AZ3166/src/libraries/WebSocket/src/WebSocketClient.cpp

```diff
@@ -0,0 +1,475 @@
+        do
+        {
+            ret = read(cmd, 200, 100);
+            if (ret < 0)
+            {
+                Serial.println("ERROR: Could not receive answer.");
+                return false;
+            }
+            cmd[ret] = '\0';
+        } while (ret > 0);
+
+        close();
+        return false;
+    }
+
+    return true;
+}
+
+bool WebSocketClient::connected()
+{
+    return ( _tcpSocket == NULL ) ? false : true;
+}
+
+int WebSocketClient::sendLength(uint32_t len, char *msg)
+{
+    if (len < 126)
+    {
+        msg[0] = len | (1 << 7);
+        return 1;
+    }
+    else if (len < 65535)
+    {
+        msg[0] = 126 | (1 << 7);
+        msg[1] = (len >> 8) & 0xff;
+        msg[2] = len & 0xff;
+        return 3;
+    }
+    else
+    {
+        msg[0] = 127 | (1 << 7);
+        for (int i = 0; i < 8; i++)
+        {
+            msg[i + 1] = (len >> i * 8) & 0xff;
+        }
+        return 9;
+    }
+}
+
+int WebSocketClient::readChar(char *pC, bool block)
+{
+    int ret = read(pC, 1, 1);
+    return ret;
+}
+
+int WebSocketClient::sendMask(char *msg)
+{
+    for (int i = 0; i < 4; i++)
+    {
+        msg[i] = 0x00;
+    }
+    return 4;
+}
+
+int WebSocketClient::send(char *str, int size)
+{
+    if (_tcpSocket == NULL)
+    {
+        Serial.println("ERROR: unable to send data when WebSocket is disconnected.");
+        return -1;
+    }
+
+    if (size <= 0)
+    {
+        Serial.println("ERROR: invalid data size.");
+        return -1;
+    }
+
+    char msg[15];
+    int idx = 1;
+    msg[0] = 0x81;
+    if (size == 0)
+    {
+        size = strlen(str);
+    }
+
+    idx += sendLength(size, msg + idx);
+    idx += sendMask(msg + idx);
+    int res = write(msg, idx);
+    if (res != idx)
+    {
+        Serial.println("ERROR: send websocket frame header failed.");
+        return -1;
+    }
+
+    res = write(str, size);
+    if (res == -1)
+    {
+        return -1;
+    }
+    else
+    {
+        return res + idx;
+    }
+}
+
+WebSocketReceiveResult *WebSocketClient::receive(char *message)
+{
+    if (_tcpSocket == NULL)
+    {
+        Serial.println("ERROR: unable to receive data when WebSocket is disconnected.");
+        return NULL;
+    }
+
+    int i = 0;
+    uint32_t payloadLength;
+    char recvByte = 0;
+    char opcode = 0;
+    char c;
+    char mask[4] = {0, 0, 0, 0};
+    bool isMasked = false;
+    bool isFinal = false;
+    Timer timer;
+
+    // Read opcode
+    timer.start();
+    while (true)
+    {
+        if (timer.read() > 10)
+        {
+            //Serial.printf("ERROR: timeout ws, timer: %d\r\n", timer.read());
+            return NULL;
+        }
+
+        int res = _tcpSocket->recv(&recvByte, 1);
+
+        if (res == 1)
+        {
+            _tcpSocket->set_timeout(2000);
+            opcode = recvByte & 0x7F;
+
+            // opcode for data frames
+            if (opcode == WS_OPCODE_CONT || opcode == WS_OPCODE_TEXT || opcode == WS_OPCODE_BINARY)

```

Comment:
if so, will ping / pong work ?

Reply:
move to TODO, we can use a small headbeat data message instead

---
In reply to: [173732540](https://github.com/Microsoft/devkit-sdk/pull/771#discussion_r173732540) [](ancestors = 173732540)